test/distributed/_composable/fsdp/test_fully_shard_extensions.py

```diff
@@ -0,0 +1,119 @@
+# Owner(s): ["oncall: distributed"]
+
+import contextlib
+import copy
+import threading
+import unittest
+from typing import Any, List, Optional, Tuple
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+
+from torch.distributed._composable.fsdp import fully_shard
+from torch.testing._internal.common_cuda import TEST_CUDA
+from torch.testing._internal.common_fsdp import (
+    check_sharded_parity,
+    FSDPTestMultiThread,
+    MLP,
+)
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.two_tensor import TwoTensor
+
+
+def two_tensor_fsdp_pre_all_gather(self):
+    all_gather_inputs = (self.a, self.b)
+    metadata = None
+    return all_gather_inputs, metadata
+
+
+def two_tensor_fsdp_post_all_gather(
+    self,
+    all_gather_outputs: Tuple[torch.Tensor, ...],
+    metadata: Any,
+    param_dtype: torch.dtype,
+    *,
+    out: Optional[torch.Tensor] = None,
+):
+    assert metadata is None, f"{metadata}"
+    a, b = all_gather_outputs
+    if out is not None:
+        assert isinstance(out, TwoTensor), f"{type(out)}"
+        assert a.untyped_storage().data_ptr() == out.a.untyped_storage().data_ptr()
+        assert b.untyped_storage().data_ptr() == out.b.untyped_storage().data_ptr()
+        return
+    tensors_to_free = (a, b)
+    return TwoTensor(a, b), tensors_to_free
+
+
+class TestFullyShardAllGatherExtensions(FSDPTestMultiThread):
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @property
+    def device(self) -> torch.device:
+        return torch.device("cuda:0")
+
+    @contextlib.contextmanager
+    def patch_two_tensor_fsdp_all_gather(self):
+        lock = threading.Lock()
```

Comment:
cc: @weifengpy Do we have any multi-threaded patching utilities? / Does this pattern look like something we already have a function for?

Reply:
nice! this is a good addon. I do not have similar utilities since I only read and write existing fields instead of deleting